src/main/java/com/google/sps/data/Room.java

```diff
@@ -14,20 +14,113 @@
     private Queue<Video> videos;
 
+    //Room factory function
+    public static Room createRoom(List<Member> members, Queue<Video> videos, Queue<Message> messages){
+        return new Room(members, videos, messages);
+    }
+
+    // Manipulates the messages property of the room given a room key and a new message to be added
+    public static void addMessagesFromKey(Key roomKey, Message chatMessage) {
+        try {
+        Entity roomEntity = datastore.get(roomKey);
+        Room room = Room.fromEntity(roomEntity);
+        room.addMessage(chatMessage);
+
+        roomEntity.setProperty("messages", room.getMessagesAsEntities());
+        datastore.put(roomEntity);
+        } catch (EntityNotFoundException e){
+            System.out.println(e.toString());
+        }
+    }
+
+    //Creates a room object from a Datastore Key
+    public static Room fromKey(long roomID) {
+        Key roomKey = KeyFactory.createKey("Room", roomID);
+        try {
+            return Room.fromEntity(datastore.get(roomKey));
+        } catch (EntityNotFoundException e) {
+            System.out.println(e.toString());
+        }
+        return null;
+    }
+
+    //Turns a Room entitiy into a Room object
+    public static Room fromEntity(Entity roomEntity) {
+        Map<String, Object> properties = roomEntity.getProperties();
+        List<Member> memberList = 
+        ((ArrayList<EmbeddedEntity>) properties.get(MEMBERS_PROPERTY)).stream().map(Member::fromEmbeddedEntity).collect(Collectors.toCollection(ArrayList::new));
+        Queue<Video> videoQueue = 
+        ((ArrayList<EmbeddedEntity>) properties.get(VIDEOS_PROPERTY)).stream().map(Video::fromEmbeddedEntity).collect(Collectors.toCollection(LinkedList::new));
+        Queue<Message> messageQueue = (Queue<Message>) properties.get(MESSAGES_PROPERTY) != null ?
+        ((ArrayList<EmbeddedEntity>) properties.get(MESSAGES_PROPERTY)).stream().map(Message::fromEmbeddedEntity).collect(Collectors.toCollection(LinkedList::new)) : new LinkedList();
+        return new Room(memberList, videoQueue, messageQueue);
+    }
+
+    //Turns the Room object into a datastore entity
+    public static Entity toEntity(Room room){
+        Entity newRoom = new Entity(ROOM_ENTITY);
+        newRoom.setProperty(MEMBERS_PROPERTY, room.getMembersAsEmbeddedEntities());
+        newRoom.setProperty(VIDEOS_PROPERTY, room.getVideosAsEntities());
+        newRoom.setProperty(MESSAGES_PROPERTY, room.getMessagesAsEntities());
+        return newRoom;
+    }
+    public static Key toDatastore(Room room){
+        Entity newRoom = Room.toEntity(room);
+        try {
+            return DatastoreServiceFactory.getDatastoreService().put(newRoom);
+        } catch (DatastoreFailureException e){
+            System.out.println(e.toString());
+        }
+        return null;
+    }
+
+    //Returns the Room's video url list
+    public Queue<Video> getVideos() {
+        return new LinkedList<Video>(this.videos);
```

Comment:
new Queue<Video>(this.videos) rather than a linkedlist since that is expected result from the function

Reply:
Queue's are interfaces, they cannot be instantiated.

Reply:
Oh, I didn't know that lol, thanks!